rfc/rbac-ng.md

```diff
@@ -0,0 +1,154 @@
+Direct use of OPA was also considered, though the flexibility enabled through using Fiat as a facade outweighed the benefits from a more direct setup.
+
+More general permission systems like SpiceDB could potentially be useful, though this would make more sense when applying permissions in a much more fine-grained fashion or to a world-wide continuous delivery service with millions of concurrent users.
+The choice of keeping Fiat as a facade does enable the potential integration with SpiceDB or other policy engines in the future, though.
+
+Clouddriver provides an API for account management.
+However, this API is based on the classes in Clouddriver.
+Reusing this API in Kayenta would require adding some form of generic account type in Clouddriver which could perform next to no validation.
+This would also complicate user secret permission enforcement.
+
+## Known Unknowns
+
+There is a lot of flexibility in using Open Policy Agent and adopting this general architecture for access controls.
+While metric store accounts have a fairly straightforward permission model, other Spinnaker resources are more complex.
+The exact API for configuring accounts and permissions may evolve.
+This should both simplify the common use case while enabling advanced use cases via access to OPA plumbing.
+
+Access controls don't make much sense without authentication, and providing authentication configurations for Kayenta beyond Spinnaker integration is out of scope for this RFC, but this doesn't prevent the addition of that in the future.
+Object store and configuration store accounts initially sounded like prime candidates for credentials management, but these could be replaced with a direct SQL backend for both objects and configurations; this is out of scope for this RFC as well.
+
+## Security, Privacy, and Compliance
+
+This feature introduces essential access controls around Kayenta metric store accounts and integration with Spinnaker authentication.
+It also depends on changes to access control checks in Gate around applications referenced in calls to Kayenta.
+
+As with Clouddriver account management, this feature works with user secrets to avoid directly storing credentials for external services.
+User secrets have their own permissions attached during secret delivery that Spinnaker uses to perform access control checks upon use.
+
+## Operations
+
+Enabling security in Kayenta requires an additional SQL database.
+If this is the first SQL database being introduced into a Spinnaker or Kayenta environment, this adds a new system to operate.
+
+The introduction of Open Policy Agent support for a Fiat backend makes fairly large changes in how Fiat is operated today.
+Instead of configuring Fiat with URLs to poll, the individual services that have permission data will be configured to push to Fiat through a common Fiat SDK.
+This should significantly reduce the chattiness and space usage of Fiat's current synchronization and polling.
```

Comment:
Do you figure the `X-SPINNAKER_ACCOUNTS` could/would disappear as part of this?

Reply:
Definitely. Realistically, the only thing we need to pass around is the userid; the rest can be offloaded to Fiat (and its own policy evaluations). However, since you can still define permissions directly on resources, we'll likely continue loading the user's roles as granted authorities to help simplify local permission checks without having to move them to policies.

Reply:
❗ 🎉 Color me excited :)